syne_tune/backend/simulator_backend/simulator_callback.py

```diff
@@ -91,6 +92,16 @@ def on_tuning_start(self, tuner: "Tuner"):
         assert tuner.sleep_time == 0, \
             "Initialize Tuner with sleep_time = 0 if you use the SimulatorBackend"
         self._time_keeper = backend.time_keeper
+        scheduler = tuner.scheduler
```

Comment:
Ok for now to add a special-case, but I propose to add a callback to avoid adding this special case (I can do that in a next PR).

Reply:
Thanks!
The callback is specific to the SimulatorBackend, which always has this time keeper.
But you are right that this testing whether scheduler is FIFOScheduler is not good, and could be improved. However, that would likely mean an extension of the Scheduler API (say, a subclass SchedulerWithTimeKeeper). So, any scheduler which also needs time, would inherit from that, and would be treated properly in the callback.

Reply:
In general, I think, the lesson here (to me) is to avoid dependencies between the components interacting in the Tuner (mainly scheduler and backend). If this cannot be avoided, these dependencies MUST be created only at the start of tuning, and not at construction.

This is because the deserialization creates a new backend object.

I think this is a good forcing function for modular design!